File: ggcommons/metrics/targets/metric_target.py
from abc import ABC, abstractmethod
import logging
from typing import Dict
from ggcommons.config.manager.config_manager import ConfigManager
from ggcommons.config.manager.configuration_change_listener import (
    ConfigurationChangeListener,
)


class MetricTarget(ConfigurationChangeListener, ABC):
    """
    Abstract base class for metric targets.
    """

    def __init__(self, config_manager: ConfigManager):
        super().__init__()
        self.config_manager = config_manager
        self.metric_config = config_manager.get_metric_config()
        self.logger = logging.getLogger(type(self).__name__)

    @abstractmethod
    def emit_metric(self, metric, measure_values: Dict[str, float]):
        """
        Abstract method to emit a metric with given measure values.
        """
        pass

    @abstractmethod
    def emit_metric_now(self, metric, measure_values: Dict[str, float]):
        """
        Abstract method to immediately emit a metric with given measure values.
        """
        pass

    @abstractmethod
    def on_configuration_change(self, configuration) -> bool:
        pass

    def close(self) -> None:
        """Release any resources held by the target (threads, files).

        No-op by default; targets that own background resources (e.g. the
        CloudWatch periodic-flush thread) override this.
        """
        pass

File: libs/java/src/main/java/com/breissinger/ggcommons/metrics/targets/MetricTarget.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.breissinger.ggcommons.metrics.targets;

import com.breissinger.ggcommons.config.ConfigManager;
import com.breissinger.ggcommons.config.ConfigurationChangeListener;
import com.breissinger.ggcommons.config.MetricConfiguration;
import com.breissinger.ggcommons.metrics.Metric;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.Map;

public abstract sealed class MetricTarget implements ConfigurationChangeListener
        permits CloudWatch, CloudWatchComponent, Messaging, Log, Prometheus
{

    protected static final Logger LOGGER = LogManager.getLogger(MetricTarget.class);

    protected final ConfigManager configManager;
    protected final MetricConfiguration metricConfig;

    MetricTarget(ConfigManager configManager)
    {
        this.configManager = configManager;
        this.metricConfig = configManager.getMetricConfig();
    }

    public abstract void emitMetric(Metric metric, Map<String, Float> measureValues);

    public abstract void emitMetricNow(Metric metric, Map<String, Float> measureValues);

    @Override
    public abstract boolean onConfigurationChanged();

    /**
     * Flushes any buffered metrics to the target. Default no-op (targets that don't buffer).
     *
     * <p><b>Lifecycle note (FR-MET-2):</b> the push targets (log/messaging/cloudwatch/
     * cloudwatchcomponent) deliver on {@code emit*}/{@code flush}. The pull-based {@link Prometheus}
     * target <em>inverts</em> this: {@code emit*} only update an in-process registry and {@code flush()}
     * is a no-op w.r.t. delivery — a scrape of {@code /metrics} pulls the current values. A caller
     * relying on {@code flush}/{@code emitMetricNow} for flush-before-exit gets nothing delivered until
     * the next scrape under the prometheus target.
     */
    public void flush() {}

    /**
     * Releases any resources held by this target (timers, clients, appenders). Default no-op. For the
     * {@link Prometheus} target this stops the embedded HTTP listener and releases its port/thread.
     */
    public void close() {}
}
